File: leetcode-04/solutions/cpp/0031.cpp
class Solution {
 public:
  void nextPermutation(vector<int>& nums) {
    int i = nums.size() - 2;

    while (i >= 0 && nums[i + 1] <= nums[i]) i--;

    if (i >= 0) {
      int j = nums.size() - 1;
      while (j >= 0 && nums[j] <= nums[i]) j--;
      swap(nums[i], nums[j]);
    }

    reverse(nums, i + 1, nums.size() - 1);
  }

 private:
  void reverse(vector<int>& nums, int l, int r) {
    while (l < r) {
      swap(nums[l], nums[r]);
      l++;
      r--;
    }
  }
};

File: leetcode-04/solutions/python3/0031.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        i = len(nums) - 2

        while i >= 0 and nums[i + 1] <= nums[i]:
            i -= 1

        if i >= 0:
            j = len(nums) - 1
            while j >= 0 and nums[j] <= nums[i]:
                j -= 1
            nums[i], nums[j] = nums[j], nums[i]

        self.reverse(nums, i + 1, len(nums) - 1)

    def reverse(self, nums: List[int], l, r) -> None:
        while l < r:
            nums[l], nums[r] = nums[r], nums[l]
            l += 1
            r -= 1
